notebooks/ch12/LR_Case_Study.md

```diff
@@ -0,0 +1,792 @@
+    '''
+    Uses gradient descent to minimize cost_fn. Returns the minimizing value of
+    theta once theta changes less than 0.001 between iterations.
+    '''
+    theta = np.zeros(X.shape[1])
+    old_cost = 10000000
+    while True:
+        if progress:
+            print(f'theta: {theta} | cost: {cost_fn(theta, X, y):.2f}')
+        gradient = grad_cost_fn(theta, X, y)
+        new_theta = theta - alpha * gradient
+        cost = cost_fn(theta, X, y)
+        if abs(old_cost - cost) < 10:
+            return new_theta
+        old_cost = cost
+        
+        theta = new_theta
+        
+"""
+        
+from scipy.optimize import minimize as sci_min
+
+def minimize(cost_fn, grad_cost_fn, X, y, progress=True):
+    '''
+    Uses scipy.minimize to minimize cost_fn using a form of gradient descent.
+    '''
+    theta = np.zeros(X.shape[1])
+    iters = 0
+    
+    def objective(theta):
+        return cost_fn(theta, X, y)
+    def gradient(theta):
+        return grad_cost_fn(theta, X, y)
+    def print_theta(theta):
+        nonlocal iters
+        if progress and iters % progress == 0:
+            print(f'theta: {theta} | cost: {cost_fn(theta, X, y):.2f}')
+        iters += 1
+        
+    print_theta(theta)
+    return sci_min(
+        objective, theta, method='BFGS', jac=gradient, callback=print_theta,
+        tol=1e-7
+    ).x
+```
+
+## Linear Regression Case Study
+
+In this section, we will perform an end-to-end case study of the application of a linear regression model to a dataset.
+
+The dataset we will be working with is a record of the population and drinking data in various states. The dataset is from this website: http://people.sc.fsu.edu/~jburkardt/datasets/regression/x20.txt, with some properties of the dataset changed for didactical purposes.
+
+Our task in this section: fit a good linear regression model to this dataset in order to predict the cirrhosis death rate.
+
+## Exploring the Dataset
+
+As in any data science project, we're going to begin by doing some exploratory data analysis.
+
+We'll start by reading in the dataset using `pandas`.
+
+
+```python
+drinking_data = pd.read_csv("drinking.csv")
+```
+
+It's always a good idea to look at _how much_ data we have by looking at the number of columns and the number of rows in our dataset.
+
+
+```python
+drinking_data.shape
+```
+
+
+
+
+    (46, 6)
+
+
+
+We have 46 rows of observations and 6 columns. The dataset is a pretty small one. Let's look at what columns we have available to us.
+
+
+```python
+drinking_data.columns.values
+```
+
+
+
+
+    array(['Bias', 'Urban population (percentage)',
+           'Late births (reciprocal * 100)', 'Wine consumption per capita',
+           'Liquor consumption per capita', 'Cirrhosis death rate'], dtype=object)
+
+
+
+But what do any of these column names mean? Looking at the documentation provided by the source of the dataset, we can create a table describing each column, sometimes referred to as a data dictionary:
```

Comment:
You can try being more explicit in this sentence that this is a common "good practice" to go to the source and look up the column descriptions, since 1 they may be unclear 2 your assumptions of what they mean can be wrong